tctl-snap/tctl-plugins/cmd/cmd.go

```diff
@@ -0,0 +1,249 @@
+package cmd
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const emailScope = "https://www.googleapis.com/auth/userinfo.email"
+
+type tokenResponse struct {
+	RefreshToken string `json:"refresh_token"`
+	AccessToken  string `json:"access_token"`
+}
+
+// ClientID returns the '<env>-google-client-id' snapctl configuration.
+func ClientID() (string, error) {
+	clientID, err := GetSnapctlArg("google-client-id")
+	if err != nil {
+		return "", err
+	}
+
+	return clientID, nil
+}
+
+// ClientSecret returns the '<env>-google-client-secret' snapctl configuration.
+func ClientSecret() (string, error) {
+	clientSecret, err := GetSnapctlArg("google-client-secret")
+	if err != nil {
+		return "", err
+	}
+
+	return clientSecret, nil
+}
+
+// FetchValidToken checks for the existence of a valid Google OAuth access token.
+// If found, it returns the token if it is not expired.
+// If expired, it refreshes the access token and returns it.
+func FetchValidToken(clientID string, clientSecret string) (string, error) {
+	path := os.Getenv("SNAP_USER_DATA")
+	accessToken, err := readTokenFromFile(path, "access")
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			fmt.Fprintf(os.Stderr, "error reading access token from file at path: %v, %v", path, err)
+			return "", err
+		}
+
+		return "", nil
```

Comment:
not sure i understand why you don't return an error here, or why do you have to return within the upper if (you could just log there no?)